### ISSUE TITLE
PID check failed. RNG must be re-initialized after fork(). Hint: Try Random.atfork()

### ISSUE THREAD

**Issue by test-account-0** [GoogleCloudPlatform/gsutil#390]
```
PID check failed. RNG must be re-initialized after fork(). Hint: Try Random.atfork()
```
gsutil exits with this message when I'm trying to copy a lot of large files to google storage:

```
 gsutil -m cp -r /mnt/some/src gs://some/dest
```
It's around 5000 files which vary in size. Their total size is around 100GB. Transfer sometimes succeeds and sometimes fails with the message above.

**Comment by thobrla** [comment]
Could save the stdout/stderr of `gsutil -D -m cp -r /mnt/some/src gs://some/dest` and paste the stack trace containing this error?

**Comment by test-account-0** [comment]
Some additional info:
```
$ gsutil version
gsutil version: 4.22
$ dpkg -l | grep python-crypto
ii  python-crypto                          2.6.1-4build1                              amd64        cryptographic algorithms and protocols for Python
$ cat /etc/issue
Ubuntu 14.04.4 LTS \n \l
```

Logs:
```
***************************** WARNING *****************************
*** You are running gsutil with debug output enabled.
*** Be aware that debug output includes authentication credentials.
*** Make sure to remove the value of the Authorization header for
*** each HTTP request printed to the console prior to posting to
*** a public medium such as a forum post or Stack Overflow.
***************************** WARNING *****************************
Command being run: /home/user/google-cloud-sdk/platform/gsutil/gsutil -o GSUtil:default_project_id=project -m -q -D rsync -r /mnt/some/src gs://some/dest
config_file_list: [...]
config: [('debug', '0'), ('working_dir', '/mnt/pyami'), ('https_validate_certificates', 'true'), ('debug', '0'), ('working_dir', '/mnt/pyami'), ('default_project_id', 'project')]
INFO 1110 11:50:48.430373 base_api.py] Calling method storage.objects.list with StorageObjectsListRequest: <StorageObjectsListRequest>

(...)

INFO 1110 12:38:11.238063 client.py] access_token is expired. Now: 2016-11-10 12:38:11.238043, token_expiry: 2016-11-10 12:35:01
PID check failed. RNG must be re-initialized after fork(). Hint: Try Random.atfork()
INFO 1110 12:38:11.240552 client.py] access_token is expired. Now: 2016-11-10 12:38:11.240539, token_expiry: 2016-11-10 12:35:01
PID check failed. RNG must be re-initialized after fork(). Hint: Try Random.atfork()

Encountered exception while syncing:
Traceback (most recent call last):
  File "/home/user/google-cloud-sdk/platform/gsutil/gslib/command.py", line 2089, in PerformTask
    results = task.func(cls, task.args, thread_state=self.thread_gsutil_api)
  File "/home/user/google-cloud-sdk/platform/gsutil/gslib/commands/rsync.py", line 1204, in _RsyncFunc
    preserve_posix=cls.preserve_posix_attrs)
  File "/home/user/google-cloud-sdk/platform/gsutil/gslib/copy_helper.py", line 3286, in PerformCopy
    allow_splitting=allow_splitting)
  File "/home/user/google-cloud-sdk/platform/gsutil/gslib/copy_helper.py", line 1795, in _UploadFileToObject
    is_component=is_component)
  File "/home/user/google-cloud-sdk/platform/gsutil/gslib/copy_helper.py", line 1588, in _UploadFileToObjectResumable
    progress_callback=progress_callback)
  File "/home/user/google-cloud-sdk/platform/gsutil/gslib/cloud_api_delegator.py", line 302, in UploadObjectResumable
    encryption_tuple=encryption_tuple, fields=fields)
  File "/home/user/google-cloud-sdk/platform/gsutil/gslib/gcs_json_api.py", line 1300, in UploadObjectResumable
    apitools_strategy=apitools_transfer.RESUMABLE_UPLOAD)
  File "/home/user/google-cloud-sdk/platform/gsutil/gslib/gcs_json_api.py", line 1135, in _UploadObject
    additional_headers, progress_callback)
  File "/home/user/google-cloud-sdk/platform/gsutil/gslib/gcs_json_api.py", line 1168, in _PerformResumableUpload
    global_params=global_params)
  File "/home/user/google-cloud-sdk/platform/gsutil/gslib/third_party/storage_apitools/storage_v1_client.py", line 1110, in Insert
    download=download)
  File "/home/user/google-cloud-sdk/platform/gsutil/third_party/apitools/apitools/base/py/base_api.py", line 687, in _RunMethod
    http_request, client=self.client)
  File "/home/user/google-cloud-sdk/platform/gsutil/third_party/apitools/apitools/base/py/transfer.py", line 838, in InitializeUpload
    retries=self.num_retries)
  File "/home/user/google-cloud-sdk/platform/gsutil/third_party/apitools/apitools/base/py/http_wrapper.py", line 351, in MakeRequest
    max_retry_wait, total_wait_sec))
  File "/home/user/google-cloud-sdk/platform/gsutil/third_party/apitools/apitools/base/py/http_wrapper.py", line 341, in MakeRequest
    check_response_func=check_response_func)
  File "/home/user/google-cloud-sdk/platform/gsutil/third_party/apitools/apitools/base/py/http_wrapper.py", line 391, in _MakeRequestNoRetry
    redirections=redirections, connection_type=connection_type)
  File "/home/user/google-cloud-sdk/platform/gsutil/third_party/oauth2client/oauth2client/client.py", line 616, in new_request
    self._refresh(request_orig)
  File "/home/user/google-cloud-sdk/platform/gsutil/third_party/oauth2client/oauth2client/client.py", line 885, in _refresh
    self._do_refresh_request(http_request)
  File "/home/user/google-cloud-sdk/platform/gsutil/third_party/oauth2client/oauth2client/client.py", line 900, in _do_refresh_request
    body = self._generate_refresh_request_body()
  File "/home/user/google-cloud-sdk/platform/gsutil/third_party/oauth2client/oauth2client/client.py", line 1611, in _generate_refresh_request_body
    assertion = self._generate_assertion()
  File "/home/user/google-cloud-sdk/platform/gsutil/third_party/oauth2client/oauth2client/service_account.py", line 396, in _generate_assertion
    key_id=self._private_key_id)
  File "/home/user/google-cloud-sdk/platform/gsutil/third_party/oauth2client/oauth2client/crypt.py", line 97, in make_signed_jwt
    signature = signer.sign(signing_input)
  File "/home/user/google-cloud-sdk/platform/gsutil/third_party/oauth2client/oauth2client/_pycrypto_crypt.py", line 101, in sign
    return PKCS1_v1_5.new(self._key).sign(SHA256.new(message))
  File "/usr/lib/python2.7/dist-packages/Crypto/Signature/PKCS1_v1_5.py", line 112, in sign
    m = self._key.decrypt(em)
  File "/usr/lib/python2.7/dist-packages/Crypto/PublicKey/RSA.py", line 174, in decrypt
    return pubkey.pubkey.decrypt(self, ciphertext)
  File "/usr/lib/python2.7/dist-packages/Crypto/PublicKey/pubkey.py", line 93, in decrypt
    plaintext=self._decrypt(ciphertext)
  File "/usr/lib/python2.7/dist-packages/Crypto/PublicKey/RSA.py", line 235, in _decrypt
    r = getRandomRange(1, self.key.n-1, randfunc=self._randfunc)
  File "/usr/lib/python2.7/dist-packages/Crypto/Util/number.py", line 123, in getRandomRange
    value = getRandomInteger(bits, randfunc)
  File "/usr/lib/python2.7/dist-packages/Crypto/Util/number.py", line 104, in getRandomInteger
    S = randfunc(N>>3)
  File "/usr/lib/python2.7/dist-packages/Crypto/Random/_UserFriendlyRNG.py", line 202, in read
    return self._singleton.read(bytes)
  File "/usr/lib/python2.7/dist-packages/Crypto/Random/_UserFriendlyRNG.py", line 178, in read
    return _UserFriendlyRNG.read(self, bytes)
  File "/usr/lib/python2.7/dist-packages/Crypto/Random/_UserFriendlyRNG.py", line 137, in read
    self._check_pid()
  File "/usr/lib/python2.7/dist-packages/Crypto/Random/_UserFriendlyRNG.py", line 153, in _check_pid
    raise AssertionError("PID check failed. RNG must be re-initialized after fork(). Hint: Try Random.atfork()")
AssertionError: PID check failed. RNG must be re-initialized after fork(). Hint: Try Random.atfork()

DEBUG: Exception stack trace:
    Traceback (most recent call last):
      File "/home/user/google-cloud-sdk/platform/gsutil/gslib/__main__.py", line 577, in _RunNamedCommandAndHandleExceptions
        collect_analytics=True)
      File "/home/user/google-cloud-sdk/platform/gsutil/gslib/command_runner.py", line 299, in RunNamedCommand
        return_code = command_inst.RunCommand()
      File "/home/user/google-cloud-sdk/platform/gsutil/gslib/commands/rsync.py", line 1431, in RunCommand
        (self.op_failure_count, plural_str, plural_str))
    CommandException: CommandException: 503 files/objects could not be copied/removed.

```

**Comment by thobrla** [comment]
Can you describe the type of credentials that you're using?  I assume you're using service account credentials based on the codepath in the stack trace.

FWIW, this looks serious but I can't reproduce it on a very similar platform.

I'm curious if this is specific to the gcloud distribution of gsutil in some way.  If you pip install gsutil in a virtualenv and then run `gsutil config -e`, pointing at the same key file, do you see a similar error?

**Comment by test-account-0** [comment]
Yes, I'm using service account credentials.

After installing gsutil in virtualenv I don't get any errors. 

pip freeze:
```
SocksiPy-branch==1.01
argparse==1.2.1
boto==2.42.0
cffi==1.9.1
crcmod==1.7
cryptography==1.5.3
enum34==1.1.6
funcsigs==1.0.2
gcs-oauth2-boto-plugin==1.14
google-apitools==0.5.3
gsutil==4.22
httplib2==0.9.2
idna==2.1
ipaddress==1.0.17
mock==2.0.0
oauth2client==2.2.0
pbr==1.10.0
pyOpenSSL==16.2.0
pyasn1==0.1.9
pyasn1-modules==0.0.8
pycparser==2.17
python-gflags==3.1.0
retry-decorator==1.1.0
rsa==3.4.2
six==1.10.0
wsgiref==0.1.2
```

**Comment by thobrla** [comment]
Thanks.  Just to be super paranoid (so we can isolate it to an issue with the gcloud bundle), can you also try downloading and extracting the standalone gsutil tarball `gs://pub/gsutil_4.22.tar.gz` and trying the same command?  This should pick up your `gsutil config -e` configuration, no need to re-run it.

**Comment by test-account-0** [comment]
I have tried and Copying have failed with the same message. Venv version is still working fine with no issues.

**Comment by thobrla** [comment]
Thanks, so it's not restricted to the gcloud bundle.  It looks like this is a problem with your Python environment since venv works without issue.  Is it possible you have an older version of apitools, oauth2client, or httplib2 or some other Python lib installed on that machine?

**Comment by test-account-0** [comment]
Yes, it is possible, almost certain.

```
# dpkg -l | grep python
ii  dh-python                              1.20140128-1ubuntu8.2                      all          Debian helper tools for packaging Python libraries and applications
ii  libboost-mpi-python-dev                1.54.0.1ubuntu1                            amd64        C++ interface to the Message Passing Interface (MPI), Python Bindings (default version)
ii  libboost-mpi-python1.54-dev            1.54.0-4ubuntu3.1                          amd64        C++ interface to the Message Passing Interface (MPI), Python Bindings
ii  libboost-mpi-python1.54.0              1.54.0-4ubuntu3.1                          amd64        C++ interface to the Message Passing Interface (MPI), Python Bindings
ii  libboost-python-dev                    1.54.0.1ubuntu1                            amd64        Boost.Python Library development files (default version)
ii  libboost-python1.54-dev:amd64          1.54.0-4ubuntu3.1                          amd64        Boost.Python Library development files
ii  libboost-python1.54.0:amd64            1.54.0-4ubuntu3.1                          amd64        Boost.Python Library
ii  libpython-dev:amd64                    2.7.5-5ubuntu3                             amd64        header files and a static library for Python (default)
ii  libpython-stdlib:amd64                 2.7.5-5ubuntu3                             amd64        interactive high-level object-oriented language (default python version)
ii  libpython2.7:amd64                     2.7.6-8ubuntu0.2                           amd64        Shared Python runtime library (version 2.7)
ii  libpython2.7-dev:amd64                 2.7.6-8ubuntu0.2                           amd64        Header files and a static library for Python (v2.7)
ii  libpython2.7-minimal:amd64             2.7.6-8ubuntu0.2                           amd64        Minimal subset of the Python language (version 2.7)
ii  libpython2.7-stdlib:amd64              2.7.6-8ubuntu0.2                           amd64        Interactive high-level object-oriented language (standard library, version 2.7)
ii  libpython3-dev:amd64                   3.4.0-0ubuntu2                             amd64        header files and a static library for Python (default)
ii  libpython3-stdlib:amd64                3.4.0-0ubuntu2                             amd64        interactive high-level object-oriented language (default python3 version)
ii  libpython3.4:amd64                     3.4.3-1ubuntu1~14.04.3                     amd64        Shared Python runtime library (version 3.4)
ii  libpython3.4-dev:amd64                 3.4.3-1ubuntu1~14.04.3                     amd64        Header files and a static library for Python (v3.4)
ii  libpython3.4-minimal:amd64             3.4.3-1ubuntu1~14.04.3                     amd64        Minimal subset of the Python language (version 3.4)
ii  libpython3.4-stdlib:amd64              3.4.3-1ubuntu1~14.04.3                     amd64        Interactive high-level object-oriented language (standard library, version 3.4)
ii  python                                 2.7.5-5ubuntu3                             amd64        interactive high-level object-oriented language (default version)
ii  python-apt                             0.9.3.5ubuntu2                             amd64        Python interface to libapt-pkg
ii  python-apt-common                      0.9.3.5ubuntu2                             all          Python interface to libapt-pkg (locales)
ii  python-bson                            2.6.3-1build1                              amd64        Python implementation of BSON for MongoDB
ii  python-bson-ext                        2.6.3-1build1                              amd64        C-coded extension to the python-bson package
ii  python-chardet                         2.0.1-2build2                              all          universal character encoding detector
ii  python-chardet-whl                     2.2.1-2~ubuntu1                            all          universal character encoding detector
ii  python-colorama                        0.2.5-0.1ubuntu2                           all          Cross-platform colored terminal text in Python - Python 2.x
ii  python-colorama-whl                    0.2.5-0.1ubuntu2                           all          Cross-platform colored terminal text in Python - Wheels
ii  python-crypto                          2.6.1-4build1                              amd64        cryptographic algorithms and protocols for Python
ii  python-debian                          0.1.21+nmu2ubuntu2                         all          Python modules to work with Debian-related data formats
ii  python-dev                             2.7.5-5ubuntu3                             amd64        header files and a static library for Python (default)
ii  python-distlib                         0.1.8-1ubuntu1                             all          low-level components of python distutils2/packaging
ii  python-distlib-whl                     0.1.8-1ubuntu1                             all          low-level components of python distutils2/packaging
ii  python-dns                             2.3.6-3                                    all          DNS client module for Python
ii  python-egenix-mxdatetime               3.2.7-1build1                              amd64        date and time handling routines for Python
ii  python-egenix-mxtools                  3.2.7-1build1                              amd64        collection of additional builtins for Python
ii  python-googleapi                       1.2-2                                      all          Google APIs Client Library for Python
ii  python-gridfs                          2.6.3-1build1                              all          Python implementation of GridFS for MongoDB
ii  python-html5lib                        0.999-3~ubuntu1                            all          HTML parser/tokenizer based on the WHATWG HTML5 specification (Python 2)
ii  python-html5lib-whl                    0.999-3~ubuntu1                            all          HTML parser/tokenizer based on the WHATWG HTML5 specification
ii  python-httplib2                        0.8-2build1                                all          comprehensive HTTP client library written for Python
ii  python-ldb                             1:1.1.24-0ubuntu0.14.04.1                  amd64        Python bindings for LDB
ii  python-meld3                           0.6.10-1                                   amd64        HTML/XML templating system for Python
ii  python-minimal                         2.7.5-5ubuntu3                             amd64        minimal subset of the Python language (default version)
ii  python-mysqldb                         1.2.3-2ubuntu1                             amd64        Python interface to MySQL
ii  python-ntdb                            1.0-2ubuntu1                               amd64        Python bindings for NTDB
ii  python-oauth2client                    1.2-2                                      all          OAuth 2.0 client library
ii  python-pcp                             3.11.6                                     amd64        Performance Co-Pilot Python PMAPI module
ii  python-pip                             1.5.4-1ubuntu3                             all          alternative Python package installer
ii  python-pip-whl                         1.5.4-1ubuntu3                             all          alternative Python package installer
ii  python-pkg-resources                   3.3-1ubuntu2                               all          Package Discovery and Resource Access using pkg_resources
ii  python-psutil                          1.2.1-1ubuntu2                             amd64        module providing convenience functions for managing processes
ii  python-psycopg2                        2.4.5-1build5                              amd64        Python module for PostgreSQL
ii  python-pycurl                          7.19.3-0ubuntu3                            amd64        Python bindings to libcurl
ii  python-pylibacl                        0.5.1-1.1build4                            amd64        module for manipulating POSIX.1e ACLs
ii  python-pymongo                         2.6.3-1build1                              amd64        Python interface to the MongoDB document-oriented database
ii  python-pymongo-ext                     2.6.3-1build1                              amd64        C-coded extension to the python-pymongo package
ii  python-pynag                           0.9.1-1461253474                           all          Python modules for Nagios plugins and configuration
ii  python-pynagio                         0.1.6-1482408540                           all          Super simple nagios check python library
ii  python-pynagioscheck                   0.2-1461247917                             all          A few functions using pynag
ii  python-pyxattr                         0.5.1-1.1build4                            amd64        module for manipulating filesystem extended attributes
ii  python-requests                        2.2.1-1ubuntu0.3                           all          elegant and simple HTTP library for Python, built for human beings
ii  python-requests-whl                    2.2.1-1ubuntu0.3                           all          elegant and simple HTTP library for Python, built for human beings
ii  python-samba                           2:4.3.9+dfsg-0ubuntu0.14.04.3              amd64        Python bindings for Samba
ii  python-scour                           0.26-3build1                               all          SVG scrubber and optimizer
ii  python-setuptools                      3.3-1ubuntu2                               all          Python Distutils Enhancements
ii  python-setuptools-whl                  3.3-1ubuntu2                               all          Python Distutils Enhancements (wheel package)
ii  python-six                             1.5.2-1ubuntu1                             all          Python 2 and 3 compatibility library (Python 2 interface)
ii  python-six-whl                         1.5.2-1ubuntu1                             all          Python 2 and 3 compatibility library (universal wheel)
ii  python-software-properties             0.92.37.7                                  all          manage the repositories that you install software from
ii  python-support                         1.0.15                                     all          automated rebuilding support for Python modules
ii  python-talloc                          2.1.5-0ubuntu0.14.04.1                     amd64        hierarchical pool based memory allocator - Python bindings
ii  python-tdb                             1.3.8-0ubuntu0.14.04.1                     amd64        Python bindings for TDB
ii  python-uritemplate                     0.5.2-1                                    all          implementation of RFC6570
ii  python-urllib3                         1.7.1-1ubuntu4                             all          HTTP library with thread-safe connection pooling for Python
ii  python-urllib3-whl                     1.7.1-1ubuntu4                             all          HTTP library with thread-safe connection pooling
ii  python-virtualenv                      1.11.4-1ubuntu1                            all          Python virtual environment creator
ii  python-wheel                           0.24.0-1~ubuntu1                           all          built-package format for Python
ii  python-xapian                          1.2.16-2ubuntu1                            amd64        Xapian search engine interface for Python
ii  python2.7                              2.7.6-8ubuntu0.2                           amd64        Interactive high-level object-oriented language (version 2.7)
ii  python2.7-dev                          2.7.6-8ubuntu0.2                           amd64        Header files and a static library for Python (v2.7)
ii  python2.7-minimal                      2.7.6-8ubuntu0.2                           amd64        Minimal subset of the Python language (version 2.7)
ii  python3                                3.4.0-0ubuntu2                             amd64        interactive high-level object-oriented language (default python3 version)
ii  python3-apt                            0.9.3.5ubuntu2                             amd64        Python 3 interface to libapt-pkg
ii  python3-chardet                        2.2.1-2~ubuntu1                            all          universal character encoding detector for Python3
ii  python3-colorama                       0.2.5-0.1ubuntu2                           all          Cross-platform colored terminal text in Python - Python 3.x
ii  python3-debian                         0.1.21+nmu2ubuntu2                         all          Python 3 modules to work with Debian-related data formats
ii  python3-dev                            3.4.0-0ubuntu2                             amd64        header files and a static library for Python (default)
ii  python3-distlib                        0.1.8-1ubuntu1                             all          low-level components of python distutils2/packaging
ii  python3-dns                            3.0.4-1ubuntu1                             all          DNS client module for Python 3
ii  python3-html5lib                       0.999-3~ubuntu1                            all          HTML parser/tokenizer based on the WHATWG HTML5 specification (Python 3)
ii  python3-minimal                        3.4.0-0ubuntu2                             amd64        minimal subset of the Python language (default python3 version)
ii  python3-pip                            1.5.4-1ubuntu3                             all          alternative Python package installer - Python 3 version of the package
ii  python3-pkg-resources                  3.3-1ubuntu2                               all          Package Discovery and Resource Access using pkg_resources
ii  python3-requests                       2.2.1-1ubuntu0.3                           all          elegant and simple HTTP library for Python3, built for human beings
ii  python3-setuptools                     3.3-1ubuntu2                               all          Python3 Distutils Enhancements
ii  python3-six                            1.5.2-1ubuntu1                             all          Python 2 and 3 compatibility library (Python 3 interface)
ii  python3-urllib3                        1.7.1-1ubuntu4                             all          HTTP library with thread-safe connection pooling for Python3
ii  python3-wheel                          0.24.0-1~ubuntu1                           all          built-package format for Python
ii  python3.4                              3.4.3-1ubuntu1~14.04.3                     amd64        Interactive high-level object-oriented language (version 3.4)
ii  python3.4-dev                          3.4.3-1ubuntu1~14.04.3                     amd64        Header files and a static library for Python (v3.4)
ii  python3.4-minimal                      3.4.3-1ubuntu1~14.04.3                     amd64        Minimal subset of the Python language (version 3.4)
```

**Comment by thobrla** [comment]
It looks like you're using oauth2client 1.2-2, gsutil requires 2.2.0

**Comment by test-account-0** [comment]
It is the newest version in ubuntu 14.04. Does gsutil still support trusty?
If it does - should I use some google repo?

**Comment by thobrla** [comment]
gsutil supports standalone installs and pip installs, but the standalone install bundles oauth2client, and that should come first in the module search path.  Do you have any idea why that might not be the case on your system?

**Comment by denis-trofimov** [comment]
Step into same error with GCS multiprocessing upload/download.
Found same problem at https://github.com/newsapps/beeswithmachineguns/issues/17

> It's a bug in paramiko, you want to add
from Crypto.Random import atfork
and atfork() at the beginning of run() in paramiko/transport.py
from @dimaqq @beeswithmachineguns

**Comment by houglum** [comment]
As I haven't seen any replies to thobrla's question on 1/4, and it looks like this is an issue with the python environment on your local system (since gsutil inside a venv works fine), I'm going to go ahead and close this.

**Comment by DollarAkshay** [comment]
I just got this issue :(

**Comment by 747project** [comment]
I just received this issue today as well

**Comment by houglum** [comment]
So, I think most folks don't see this because it happens under a subset of runtime circumstances.  We use oauth2client, which has mechanisms in place to fall back to "pure Python crypto" in the absence of pyOpenSSL -- additionally, I believe this only happens when using service account credentials.  Fun.

From the [PyCrypto](https://github.com/dlitz/pycrypto) docs, I think the fix here is to call `Crypto`'s `Random.atfork()` method (if `Crypto` is installed and can be imported) in the parent and child every time we spawn a new process, whether that be with `os.fork()`, `multiprocessing.Process()`, etc.:

> Caveat: For the random number generator to work correctly, you must
call Random.atfork() in both the parent and child processes after
using os.fork()

**Comment by houglum** [comment]
The fix will be included in the next version of gsutil (v4.36), whenever that is released.  (I don't currently have an ETA for this.)

**Comment by jmcabandara** [comment]
I got the error as follows.

`Failed RNG.`
